Fix failure when reading a table having partition columns renamed by Spark

Hi, Trino community,

I found this issue when I tried the renaming feature of Delta Lake. It appeared that Trino did not support this feature weeks ago, and it started to support this in the recent update. However, I found when I tried to rename the partition columns, the query would still fail while the query on SparkSQL is correct. (It works fine for non-partition columns).

This test is to rename the partition "department" to "dep"
![Screen Shot 2022-08-05 at 11 22 08 AM](https://user-images.githubusercontent.com/24840232/183138206-a069a771-6baf-44ff-9c1a-bc33885cbf77.png)

I could reproduce the issue using the below steps. Let us know if it's different from your intention. 

```sql
-- Databricks
CREATE TABLE test_rename (c1 int, c2 int) PARTITIONED BY (c2) TBLPROPERTIES ('delta.columnMapping.mode' = 'name', 'delta.minReaderVersion' = '2', 'delta.minWriterVersion' = '5');
INSERT INTO test_rename VALUES (1, 1);
ALTER TABLE test_rename RENAME COLUMN c2 TO new_c2;

-- Trino
SELECT * FROM test_rename;
io.trino.spi.TrinoException: Index 1 out of bounds for length 1
	at io.trino.plugin.deltalake.DeltaLakePageSource.getNextPage(DeltaLakePageSource.java:176)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:311)
	at io.trino.operator.Driver.processInternal(Driver.java:410)
	at io.trino.operator.Driver.lambda$process$10(Driver.java:313)
	at io.trino.operator.Driver.tryWithLock(Driver.java:698)
	at io.trino.operator.Driver.process(Driver.java:305)
	at io.trino.operator.Driver.processForDuration(Driver.java:276)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:740)
	at io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:164)
	at io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:490)
	at io.trino.$gen.Trino_dev____20220808_024050_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at io.trino.spi.Page.getBlock(Page.java:131)
	at io.trino.plugin.deltalake.DeltaLakePageSource.getNextPage(DeltaLakePageSource.java:168)
	... 13 more
```

--- a/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSource.java
+++ b/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSource.java
@@ -85,9 +85,9 @@ public class DeltaLakePageSource
         Block partitionsBlock = null;
 
         for (DeltaLakeColumnHandle column : columns) {
-            if (partitionKeys.containsKey(column.getName())) {
+            if (partitionKeys.containsKey(column.getPhysicalName())) {
                 Type type = column.getType();
-                Object prefilledValue = deserializePartitionValue(column, partitionKeys.get(column.getName()));
+                Object prefilledValue = deserializePartitionValue(column, partitionKeys.get(column.getPhysicalName()));
                 prefilledBlocks[outputIndex] = Utils.nativeValueToBlock(type, prefilledValue);
                 delegateIndexes[outputIndex] = -1;
             }

--- a/testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeAlterTableCompatibility.java
+++ b/testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeAlterTableCompatibility.java
@@ -23,6 +23,7 @@ import static io.trino.tempto.assertions.QueryAssert.assertQueryFailure;
 import static io.trino.tempto.assertions.QueryAssert.assertThat;
 import static io.trino.tests.product.TestGroups.DELTA_LAKE_DATABRICKS;
 import static io.trino.tests.product.TestGroups.DELTA_LAKE_EXCLUDE_73;
+import static io.trino.tests.product.TestGroups.DELTA_LAKE_EXCLUDE_91;
 import static io.trino.tests.product.TestGroups.DELTA_LAKE_OSS;
 import static io.trino.tests.product.TestGroups.PROFILE_SPECIFIC_TESTS;
 import static io.trino.tests.product.deltalake.util.DeltaLakeTestUtils.getColumnCommentOnDelta;
@@ -76,6 +77,71 @@ public class TestDeltaLakeAlterTableCompatibility
         try {
             assertQueryFailure(() -> onTrino().executeQuery("ALTER TABLE delta.default." + tableName + " ADD COLUMN new_col int"))
                     .hasMessageMatching(".* Table .* requires Delta Lake writer version 3 which is not supported");
+        }
+        finally {
+            onDelta().executeQuery("DROP TABLE default." + tableName);
+        }
+    }
+
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_73, DELTA_LAKE_EXCLUDE_91, PROFILE_SPECIFIC_TESTS})
+    public void testRenameColumn()
+    {
+        String tableName = "test_dl_rename_column_" + randomTableSuffix();
+        String tableDirectory = "databricks-compatibility-test-" + tableName;
+
+        onDelta().executeQuery(format("" +
+                        "CREATE TABLE default.%s (col INT) " +
+                        "USING DELTA LOCATION 's3://%s/%s' " +
+                        "TBLPROPERTIES ('delta.columnMapping.mode'='name')",
+                tableName,
+                bucketName,
+                tableDirectory));
+
+        try {
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES (1)");
+            assertThat(onTrino().executeQuery("SELECT col FROM delta.default." + tableName))
+                    .containsOnly(row(1));
+
+            onDelta().executeQuery("ALTER TABLE default." + tableName + " RENAME COLUMN col TO new_col");
+            assertThat(onTrino().executeQuery("SELECT new_col FROM delta.default." + tableName))
+                    .containsOnly(row(1));
+
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES (2)");
+            assertThat(onTrino().executeQuery("SELECT new_col FROM delta.default." + tableName))
+                    .containsOnly(row(1), row(2));
+        }
+        finally {
+            onDelta().executeQuery("DROP TABLE default." + tableName);
+        }
+    }
+
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_73, DELTA_LAKE_EXCLUDE_91, PROFILE_SPECIFIC_TESTS})
+    public void testRenamePartitionedColumn()
+    {
+        String tableName = "test_dl_rename_partitioned_column_" + randomTableSuffix();
+        String tableDirectory = "databricks-compatibility-test-" + tableName;
+
+        onDelta().executeQuery(format("" +
+                        "CREATE TABLE default.%s (col INT, part STRING) " +
+                        "USING DELTA LOCATION 's3://%s/%s' " +
+                        "PARTITIONED BY (part) " +
+                        "TBLPROPERTIES ('delta.columnMapping.mode'='name')",
+                tableName,
+                bucketName,
+                tableDirectory));
+
+        try {
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES (1, 'part1')");
+            assertThat(onTrino().executeQuery("SELECT col, part FROM delta.default." + tableName))
+                    .containsOnly(row(1, "part1"));
+
+            onDelta().executeQuery("ALTER TABLE default." + tableName + " RENAME COLUMN part TO new_part");
+            assertThat(onTrino().executeQuery("SELECT col, new_part FROM delta.default." + tableName))
+                    .containsOnly(row(1, "part1"));
+
+            onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES (2, 'part2')");
+            assertThat(onTrino().executeQuery("SELECT col, new_part FROM delta.default." + tableName))
+                    .containsOnly(row(1, "part1"), row(2, "part2"));
         }
         finally {
             onDelta().executeQuery("DROP TABLE default." + tableName);

